subsys/net/ip/tcp.c

```diff
@@ -2639,124 +2671,221 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 	case TCP_CLOSED:
 		break;
 	case TCP_FIN_WAIT_1:
+		/*
+		 * FIN1:
+		 * Acknowledge path and sequence path are independent, treat them that way
+		 * The table of incoming messages and their destination states:
+		 * -   & -   -> TCP_FIN_WAIT_1
+		 * FIN & -   -> TCP_CLOSING
+		 * -   & ACK -> TCP_FIN_WAIT_2
+		 * FIN & ACK -> TCP_TIME_WAIT
+		 */
 		if (th) {
-			/* Acknowledge but drop any new data */
-			if (len > 0) {
-				int32_t new_len = len - net_tcp_seq_cmp(conn->ack, th_seq(th));
-				/* Cases:
-				 * - Data already received earlier: len > 0 , new_len <= 0
-				 * - Partially new data len > 0, new_len > 0
-				 * - Out of order data len > 0, new_len > 0, ignore the data in
-				 *   between
+			bool fin_acked = false;
+
+			if (tcp_compute_new_length(conn, th, len, false) > 0) {
+				/* We do not implement half closed sockets, therefore
+				 * cannot accept new data in after sending our FIN, as
+				 * we are in sequence can send a reset now.
 				 */
-				if (new_len > 0) {
-					conn_ack(conn, + new_len);
-				}
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
```

Comment:
Could you add connection identifier here so that it is possible to debug things if there are multiple streams going on at the same time. So something like `conn %p` like in other debug prints in this file.

Reply:
Good suggestion